source/elements/oneDAL/source/algorithms/clustering/kmeans.rst

```diff
@@ -1,153 +1,494 @@
+
+      Getter
+         | ``const table& get_centroids() const``
+
+
+   .. member:: std::int64_t cluster_count = 0
 
-Training
---------
+      Number of clusters :math:`k` in the trained model.
 
+      Getter
+         | ``std::int64_t get_cluster_count() const``
+
+      Invariants
+         | :expr:`cluster_count == centroids.row_count`
+
+
+Training :expr:`onedal::train(...)`
+-----------------------------------
 Input
 ~~~~~
-.. code-block:: c++
-
-   namespace onedal::kmeans {
+::
 
    class train_input {
    public:
       train_input();
       train_input(const table& data);
       train_input(const table& data, const table& initial_centroids);
 
+      const table& get_data() const;
+      const table& get_initial_centroids() const;
+
       train_input& set_data(const table&);
       train_input& set_initial_centroids(const table&);
    };
 
-   } // namespace onedal::kmeans
+.. class:: train_input
+
+   .. function:: train_input()
+
+      Creates input for the training operation with the default attribute
+      values.
+
+
+   .. function:: train_input(const table& data)
+
+      Creates input for the training operation with the given :expr:`data`, the
+      other attributes get default values.
+
+
+   .. function:: train_input(const table& data, const table& initial_centroids)
+
+      Creates input for the training operation with the given data and
+      :expr:`initial_centroids`.
+
+
+   .. member:: table data = table()
+
+      :math:`n \times p` table with the data to be clustered, where each row
+      stores one sample.
+
+      Getter & Setter
+         | ``const table& get_data() const``
+         | ``train_input& set_data(const table&)``
+
+
+   .. member:: table initial_centroids = table()
+
+      :math:`k \times p` table with the initial centroids, where each row
+      stores one centroid.
+
+      Getter & Setter
+         | ``const table& get_initial_centroids() const``
+         | ``train_input& set_initial_centroids(const table&)``
 
 
 Result
 ~~~~~~
-.. code-block:: c++
-
-   namespace onedal::kmeans {
+::
 
    class train_result {
    public:
+      train_result();
+
       const model& get_model() const;
       const table& get_labels() const;
-      std::int64_t get_iteration_count() const;
+      int64_t get_iteration_count() const;
       double get_objective_function_value() const;
    };
 
-   } // namespace onedal::kmeans
+.. class:: train_result
 
+   .. function:: train_result()
 
-Usage example
-~~~~~~~~~~~~~
+      Creates result of the training operation with the default attribute
+      values.
 
-.. code-block:: c++
 
-   int main() {
-      auto kmeans_desc = onedal::kmeans::desc{}
-         .set_cluster_count(10);
+   .. member:: kmeans::model model = kmeans::model()
+
+      The trained K-means model.
+
+      Getter
+         | ``const model& get_model() const``
+
+
+   .. member:: table labels = table()
+
+      :math:`n \times 1` table with assignments of cluster indices to samples in
```

Comment:
is it samples or observations?

Reply:
I'd like to change terminology, observation/feature vector -> sample

Reply:
I'l update corresponding sections

Reply:
looking from the perspective of algorithms and tables with features/observations in 0.7, I'd decide on it a bit later

Reply:
ok